File: Algorithms/127_WordLadder.java
public class Solution {
      public int ladderLength(String start, String end, Set<String> wordDict) {
        if (start.equals(end)) {
            return 2;
        }
        LinkedList<String> q = new LinkedList<String>();
        q.push(start);
        Set<String> visited = new HashSet<String>();
        visited.add(start);
        Map<String, String> m = new HashMap<String, String>();
        while (!q.isEmpty()) {
            String tmp = q.remove();
            for (int i=0; i < tmp.length(); i++) {
                for (char c='a'; c <= 'z'; c++) {
                    if (c == tmp.charAt(i)) continue;
                    char[] t = tmp.toCharArray();
                    t[i] = c;
                    String temp = new String(t);
                    if (temp.equals(end)) {
                        int length = 2;
                        while (!tmp.equals(start)) {
                            tmp = m.get(tmp);
                            length ++;
                        }
                        return length;
                    }
                    if (wordDict.contains(temp) &&
                            !visited.contains(temp)) {
                        q.add(temp);
                        visited.add(temp);
                        m.put(temp, tmp);
                    }
                }
            }
        }
        return 0;
    }
}

File: 127.WordLadder.cpp
class Solution {
public:
    int ladderLength(string start, string end, unordered_set<string> &dict) {
        if (start == end) {
            return 2;
        }
        queue<string> q;
        q.push(start);
        unordered_set<string> visited;
        visited.insert(start);
        map<string, string> m;
        while (!q.empty()) {
            string tmp = q.front();
            q.pop();
            for (int i=0; i < tmp.size(); i++) {
                for (char c='a'; c <= 'z'; c++) {
                    if (c == tmp[i]) continue;
                    string temp = tmp;
                    temp[i] = c;
                    if (temp == end) {
                        int length = 2;
                        while (tmp != start) {
                            tmp = m[tmp];
                            length++;
                        }
                        return length;
                    }
                    
                    if (dict.find(temp) != dict.end() && 
                            visited.find(temp) == visited.end()) {
                        q.push(temp);
                        visited.insert(temp);
                        m[temp] = tmp;
                    }
                }
            }
        }
        return 0;
    }
};
